### PR TITLE
Another round of API clean-up

### PR DESCRIPTION
**Pull request by trustin**
- Removed all deprecated classes, fields and methods from the public
  API.
- The `-Dcom.linecorp.armeria.useLegacyMeterNames` flag does not work
  anymore.
- Removed `HTTP_` or `Http` from various fields and methods.
  - `ClientOptions`:
    - `HTTP_HEADERS` -> `HEADERS`
    - `httpHeaders()` -> `headers()`
  - `Clients`:
    - `withHttpHeader()` -> `withHeader()`
    - `withHttpHeaders()` -> `withHeaders()`
  - `AbstractClientOptionsBuilder` and all its subtypes
    - `setHttpHeader()` -> `setHeader()`
    - `setHttpHeaders()` -> `setHeaders()`
    - `addHttpHeader()` -> `addHeader()`
    - `addHttpHeaders()` -> `addHeaders()`
- Removed redundant `add` prefix to some builder methods.
  - `RequestContextExporterBuilder`
    - `addBuiltIn()` -> `builtIn()`
    - `addAttribute()` -> `attr()`
    - `addRequestHeader()` -> `requestHeader()`
    - `addResponseHeader()` -> `responseHeader()`
    - `addKeyPattern()` -> `keyPattern()`
  - `KeyStoreCredentialResolverBuilder`
    - `addKeyPassword()` -> `keyPassword()`
    - `addKeyPasswords()` -> `keyPasswords()`
- `DocService`, `DocServiceBuilder`, and their related things
  - `exampleHttpHeaders` -> `exampleHeaders`
  - `exampleRequest()` -> `exampleRequests()`
  - `exampleRequestForMethod()` -> `exampleRequests()`
    - Removed `exampleRequests(Object...)` to avoid ambiguity.
  - `injectedScript()` -> `injectedScripts()`
  - Removed `http_` from all query parameters
    - `http_headers` -> `headers`
    - `http_headers_sticky` -> `sticky_headers`
- `AbstractRequestContextBuilder` and all its subtypes
  - `remoteAddress()` and `localAddress()` now accept `SocketAddress`
    instead of `InetSocketAddress`, so that we have more options when we
    add other transports such as Unix domain socket.
- `ServerListenerBuilder`
  - `addStartingCallback()` -> `whenStarting()`
  - `addStartedCallback()` -> `whenStarted()`
  - `addStopingCallback()` -> `whenStoping()`
  - `addStopedCallback()` -> `whenStoped()`
  - Removed all methods that accept `Runnable` in favor of those that
    accept `Consumer`.
  - Fixed the type parameters of `Iterable`s.
- `EncodingServiceBuilder`
  - `encodableContentTypePredicate()` -> `encodableContentTypes()`
  - `encodableRequestHeadersPredicate()` -> `encodableRequestHeaders()`
- `TokenBucketThrottlingStrategyBuilder` /cc @max904-github
  - Removed the `with` from the builder methods.
    - `withMinimumBackoff()` -> `minimumBackoff()`
    - `withHeadersScheme()` -> `headersScheme()`
- Miscellaneous:
  - Precalculate the hash code of `DefaultRoute`.
  - Allow binding `ServiceWithRoutes` without specifying any routes.
    - This allows a user to bind a `GrpcService` without binding any routes explicitly.

### CODE REVIEW

**Comment by trustin** [comment]
Let me check if the front-end changes didn't break anything tomorrow and remove the draft tag.

File: core/src/main/java/com/linecorp/armeria/client/endpoint/healthcheck/HealthCheckedEndpointGroup.java [hunk 1]
```diff
@@ -231,8 +230,7 @@ protected void doCloseAsync(CompletableFuture<?> future) {
      * {@link HealthCheckedEndpointGroup} with the default meter names.
      */
     public MeterBinder newMeterBinder(String groupName) {
-        return newMeterBinder(new MeterIdPrefix(Flags.useLegacyMeterNames() ? "armeria.client.endpointGroup"
-                                                                            : "armeria.client.endpoint.group",
+        return newMeterBinder(new MeterIdPrefix("armeria.client.endpoint.group",
                                                 "name", groupName));
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/client/endpoint/healthcheck/HealthCheckedEndpointGroup.java:234]: nit: could merge these two lines?

File: core/src/main/java/com/linecorp/armeria/server/Server.java [hunk 2]
```diff
@@ -486,9 +481,8 @@ private void setupServerMetrics() {
             final GracefulShutdownSupport gracefulShutdownSupport = this.gracefulShutdownSupport;
             assert gracefulShutdownSupport != null;
 
-            meterRegistry.gauge(Flags.useLegacyMeterNames() ? "armeria.server.pendingResponses"
-                                                            : "armeria.server.pending.responses",
-                                gracefulShutdownSupport, GracefulShutdownSupport::pendingResponses);
+            meterRegistry.gauge("armeria.server.pending.responses",gracefulShutdownSupport,
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/server/Server.java:484]: nit: need a space after `,`

File: core/src/main/java/com/linecorp/armeria/server/docs/DocService.java [hunk 3]
```diff
@@ -331,13 +331,13 @@ private ServiceInfo addServiceExamples(ServiceInfo service) {
...
-                        concatAndDedup(exampleHttpHeaders.get(m.name()), m.exampleHttpHeaders()),
+                        concatAndDedup(exampleHeaders.get(m.name()), m.exampleHeaders()),
                         concatAndDedup(exampleRequests.get(m.name()), m.exampleRequests()),
                         examplePaths.get(m.name()),
                         exampleQueries.get(m.name()),
                         m.httpMethod(), m.docString()))::iterator,
-                Iterables.concat(service.exampleHttpHeaders(),
-                                 exampleHttpHeaders.get("")),
+                Iterables.concat(service.exampleHeaders(),
+                                 exampleHeaders.get("")),
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/server/docs/DocService.java:340]: nit: Could merge into one line?

File: core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreakerListener.java [hunk 4]
```diff
@@ -31,15 +26,84 @@
...
+     *   <td>The number of circuit breaker state transitions to {@link CircuitState#OPEN}.</td>
+     * </tr>
+     * <tr>
+     *   <td>
+     *     {@code armeria.client.circuit.breaker.transitions{name="<circuit breaker name>",state="HALF_OPEN"}}
+     *   </td>
+     *   <td>The number of circuit breaker state transitions to {@link CircuitState#HALF_OPEN}.</td>
+     * </tr>
+     * <tr>
+     *   <td>{@code armeria.client.circuit.breaker.rejectedRequests{name="<circuit breaker name>"}}</td>
```
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreakerListener.java:64]: Ditto

**Comment by trustin** [comment]
Thanks a lot for quick reviews 😄

**Pull request merged**